### ISSUE TITLE
Review page getting stuck loading

### ISSUE THREAD

**Issue by magnonellie** [fabd/kanji-koohii#104]
When on an unstable connection, the review process may lock up and be stuck until you force refresh, causing you to lose a lot of progress you made.

**fabd**: the error in question:

> Uncaught TypeError: Cannot set property 'innerHTML' of undefined

## How to reproduce
It's fairly simple to reproduce this bug with chrome devtools.
- Start a review session
- Open devtools
- Go to the network tab
- Under throttling, add a custom profile with 1kb/s download and upload with a lot (about 100000 should do) of latency
- Review until it tries to sync progress with the server
- Wait a few seconds and you should see an error popping up in the console, seemingly spawned from YUI trying to create an error dialog.

## Possible cause
I think it's pretty clear just from how you can reproduce this that it's a timeout that's set somewhere which doesn't handle errors in a proper way.

## Possible fix
The easiest fix is probably gonna be to just catch the error and delay executing another ping for a while, still catching any errors the other ping may introduce. While we could try to listen for online / offline events, these are often unreliable since a situation like the one we made up with throttling often occurs in reality with mobile connections, causing these events to not fire correctly since they think that there still technically is a connection.
Something else that might be considered is increasing the amount of cards that's stored on the reviewing device at a time. There is way more than plenty of RAM on any semi-modern device today, so that shouldn't pose a problem at all. Doing this would only partially fix the problem and the solution above would still be required. It would however provide a much better experience for people traveling on trains for example.

**Comment by fabd** [comment]
> Something else that might be considered is increasing the amount of cards that's stored on the reviewing device at a time.

We can in theory just put ALL the cards in the page. This is purposefully kept to 10 so that user never loses too many reviews. On mobile device especially, the app could be unloaded at any time to free memory if it's not in the foreground. The user would lose potentially 100+ reviews on their commute, say.

There's always solutions of course, but in this legacy code I don't think it's worth the trouble. One solution would be to store everything in localStorage or similar. If the app is killed, and restarted, the state could be restored. And then eventually, sync to server.

**Comment by magnonellie** [comment]
That's true. I didn't actually think about that. I haven't looked too much into how you're handling the flow here, but couldn't you download all the cards and then just push progress every now and then without fetching new cards?

**Comment by fabd** [comment]
**There is another reason we don't buffer all the cards**

Review modes with example words require additional queries for every flashcard. These are not trivial to optimize. So if we buffered 100 cards, we would have to run these queries on 100 cards, which is a lot to do at once. Buffering cards in small batches means we can do more complex queries for example vocab and such, as it doesn't happen all at once.

**Comment by fabd** [comment]
> couldn't you download all the cards and then just push progress every now and then without fetching new cards?

Where do you download them? You are in a sandboxed javascript environment. You can't save a file (bar the local storage option I mentioned). So all this data goes client side to the Javascript memory. If the app dies, or is unloaded to make space for other apps on the user's device.. (ie. look at how routinely Safari / Chrome tabs get reloaded). .. then all this data / session goes POOF.

**Comment by magnonellie** [comment]
I don't know why you're mad at me, but I did mean to localstorage which seems perfect for this purpose. :smile:

**Comment by fabd** [comment]
Hmm? Just replied at the same time. Having a look at it now.

**Comment by fabd** [comment]
For reference this may be useful , this is what shows in development:

![screen shot 2017-04-29 at 14 26 27](https://cloud.githubusercontent.com/assets/169391/25555487/f3a998cc-2ce7-11e7-814d-f0a6807913ec.png)

**Comment by fabd** [comment]
First we see `kanji-flashcardreview.juicy.js` setErrorDialog() method.

`.juicy.js` is the custom Juicer file naming convention, meaning this file can include many other files with a modular approach.

Likely the element AjaxQueue expects to display an error message is not passed correctly, or the reference is lost after some updates to the view templates.

**Comment by fabd** [comment]
Another thing that is good to know, is Juicer provides a form of **hot reload** through Apache mod_rewrite.

Therefore, if we view source, and click through this file, we get the full build of the file, with all its "Juicer" includes, and all the Javascript comments. Basically production minus minifying.

`http://koohii.local/version/cache.php?env=dev&app=koohii&path=%2Frevtk%2Fkanji-flashcardreview.juicy.js`

**cache.php** is the script which parses a .juicer files.

TLDR: it's super easy to find the code which breaks, development has a lot more info.

**Comment by fabd** [comment]
Here is the elusive animal:

![85uaaaaabjru5erkjggg](https://cloud.githubusercontent.com/assets/169391/25555609/8b723634-2ceb-11e7-88e0-e767c443553d.png)

Next I will explain how I went through this (relatively simple) bug as it could help contributors.

**Comment by fabd** [comment]
I will break down a simple bug hunt. Hopefully shows while the codebase is old, it's really not that hard to debug.

1. Redproducing the error, thanks to ellie (Chrome network tab > throttle request).
2. Flip a bunch of cards (rapidly alternate N & spacebar presses) till it tries to load an ew batch of cards.
3. See the following in the console.

![f3a998cc-2ce7-11e7-814d-f0a6807913ec](https://cloud.githubusercontent.com/assets/169391/25555743/bde2b226-2cee-11e7-969a-91df5d460856.png)

Of note:

- we have a lot of development debug info we can keep out of production with `Core.log()` calls.
- we know where it breaks, and what files are concerned

4. Admittedly here knowing how the Flashcard Review front end is bundled helps. But we can trace this:

4a. We can fuzzyfind AjaxQueue class in text editor. Look at setErrorDialog(). We can see it searches for a child element of class "msg". 

Better yet the file is well documented  and confirms:

```
 *   elError    {String|HTMLElement}  Dialog displayed when a connection error occured (anything other
 *                                    than HTTP 200. Element style display will be set to 'block'.
 *                                    Must contain child element with class "uiFcAjaxError_msg" which will be
 *                                    set to the error message. Must contain a link that will enable
 *                                    use to reconnect (retry the last request, continue the queue).
```

4b. Can do a few things here. Could start by trying to locate this "msg" in a view template with a grep. I have a grep alias `grepcd '"msg"'. Typically the class would be inside double quotes. Already we have a problem here: this old code uses a way-too-generic class name, hard to search for.

4c Or we can grep for AjaxQueue, see how it is initialized in [https://github.com/fabd/kanji-koohii/blob/master/web/revtk/classes/FlashcardReview.js#L203](FlashcardReview.js)

This reveals two elements that should be in our template: `uiFcAjaxError`and `uiFcAjaxLoading`:

```
      // init ajax
      this.ajaxQueue = new Core.Ui.AjaxQueue({
        elError:    'uiFcAjaxError', 
        elLoading:  'uiFcAjaxLoading',
        events:     {
          onSuccess: Core.bind(this.onAjaxSuccess, this)
        }
      });
```

So let('s grep for that, and we trace our way back to the view templates:

```
./apps/koohii/modules/labs/templates/_iTouchVocabReview.php:7:<div id="uiFcAjaxError" style="display:none">
./apps/koohii/modules/labs/templates/_iTouchVocabReview.php:8:  <span class="uiFcAjaxError_msg">Oops!</span>&nbsp;&nbsp;<a href="#">Reconnect</a>
./apps/koohii/modules/review/templates/_ReviewKanji.php:7:<div id="uiFcAjaxError" style="display:none">
./apps/koohii/modules/review/templates/_ReviewKanji.php:8:  <span class="uiFcAjaxError_msg">Oops!</span>&nbsp;&nbsp;<a href="#">Reconnect</a>
./web/revtk/components/MobileFlashcardReview.css:28:#uiFcAjaxError { height:32px; position:absolute; right:22px; top:3px; z-index:1000; }
./web/revtk/classes/FlashcardReview.js:93: *   #uiFcAjaxError  
./web/revtk/classes/FlashcardReview.js:205:        elError:    'uiFcAjaxError', 
./web/revtk/bundles/flashcardreview-1.0.juicy.css:35:#uiFcAjaxError { 
./web/revtk/bundles/flashcardreview-1.0.juicy.css:40:#uiFcAjaxError a { color:#fff; font-weight:bold; }
./lib/front/corejs/ui/ajaxqueue.js:30: *                                    Must contain child element with class "uiFcAjaxError_msg" which will be
./lib/front/corejs/ui/ajaxqueue.js:285:        var el = Dom.down(this.elAjaxError, 'uiFcAjaxError_msg');
```
5, Open the Flashcard Review templates (php): `/apps/koohii/modules/labs/templates/_iTouchVocabReview.php` and indeed we find the concerned elements. We can see they are hidden until needed:

```html
<?php # Ajax loading indicator ?>
<div id="uiFcAjaxLoading" style="display:none"><span class="spinner"></span>Loading</div>

<?php # Connection timeout message ?>
<div id="uiFcAjaxError" class="uiFcErrorMsg" style="display:none">
  Oops!&nbsp;&nbsp;<a href="#">Reconnect</a>
</div>
```

So we notice two things:

* There is a class `uiFcErrorMsg` that seems related to this "msg" class...
* ... but where is the child element of "msg" class?

Conclusion: maybe something was lots in translation during refactoring.

6a. First we don't want to break something so let's grep for `uiFcErrorMsg`is it used anywhere? No.

6b. Then we conclude it was probably a broken refactor. Remove this class,and add a new child element, but while we are at it, let's change `msg` into a more unique class name `uiFcAjaxError_msg`.. So when we look at this months down the line, it's super easy to find the concerned code with a text search:

```html
<?php # Ajax loading indicator ?>
<div id="uiFcAjaxLoading" style="display:none"><span class="spinner"></span>Loading</div>

<?php # Connection timeout message ?>
<div id="uiFcAjaxError" style="display:none">
  <span class="uiFcAjaxError_msg">Oops!</span>&nbsp;&nbsp;<a href="#">Reconnect</a>
</div>
```

7. Then obviously we update AjaxQueue.js to use this new class.
7b. And let's not forget there were two templates concerned, revealed by our first search: `_iTouchVocabReview.php`and `_ReviewKanji.php`. Oddly named files... the first once is the vocab shuffle review modes. The other one is all the kanji flashcard reviews. It's not ideal but the reason why there are two templates (and some amount of DRY) is because the vocab cards have a different format. If I wrote this today or a couple years ago, it may have been better... but Vocab Cards were an after thought. And at that point, it made more sense that managing two relatively similar view templates would be less of a headache, then adding more if / else  cases into a singular templates (the KanjiReview template already handles quite a few review modes and cases like example words).

And that's it. A really silly bug.
